File: pyscripts/fetch_tournaments.py
#!/usr/bin/env python3
"""
Fetch the official WTA Tour calendar for the active seasons.

The tournament feed carries the full draw metadata plus the champion of each
completed event, which is what the calendar view renders.
"""

from __future__ import annotations

import re

from datetime import datetime, timezone

from wtalib import env_int, fetch_tournaments, log, write_json

YEAR = env_int("WTA_SEASON", datetime.now(timezone.utc).year)
PREV = YEAR - 1

# Main-tour levels only: ITF World Tennis Tour events are out of scope.
TOUR_LEVELS = {
    "Grand Slam",
    "WTA Finals",
    "WTA 1000",
    "WTA 500",
    "WTA 250",
    "WTA 125",
    "United Cup",
    "Billie Jean King Cup",
}

LEVEL_ORDER = [
    "Grand Slam",
    "WTA Finals",
    "WTA 1000",
    "WTA 500",
    "WTA 250",
    "WTA 125",
    "ITF",
    "United Cup",
    "Billie Jean King Cup",
]


def title_case(value: str) -> str:
    """
    Match the reference implementation's casing exactly.

    It splits on whitespace and rejoins with single spaces *without* stripping,
    so a label such as "Rome " keeps its trailing space.  Trimming here would
    make the snapshot differ from the one it replaces, so the behaviour is kept.
    """
    parts = re.split(r"\s+", str(value).lower())
    return " ".join(w[:1].upper() + w[1:] if w else w for w in parts)


def collect(year: int) -> list[dict]:
    out: list[dict] = []
    for page in range(12):
        res = fetch_tournaments(
            {"page": page, "pageSize": 100, "from": f"{year}-01-01", "to": f"{year}-12-31"}
        )
        content = (res or {}).get("content")
        if not isinstance(content, list) or not content:
            break
        out.extend(
            t
            for t in content
            if (t.get("level") or (t.get("tournamentGroup") or {}).get("level")) in TOUR_LEVELS
        )
        if len(content) < 100:
            break
    return out


def normalise(tournament: dict) -> dict:
    group = tournament.get("tournamentGroup") or {}
    winners = tournament.get("winners") or []
    winner = winners[0] if winners else None
    singles = ((winner or {}).get("singles") or {}).get("player")
    doubles = (winner or {}).get("doubles")
    doubles = doubles if isinstance(doubles, list) else []

    return {
        "id": group.get("id"),
        "name": title_case(group.get("name") or tournament.get("title") or ""),
        "title": tournament.get("title") or "",
        "year": tournament.get("year"),
        "level": tournament.get("level") or group.get("level") or "",
        "start": str(tournament.get("startDate") or "")[:10],
        "end": str(tournament.get("endDate") or "")[:10],
        "surface": tournament.get("surface") or "",
        "indoor": tournament.get("inOutdoor") == "I",
        "city": title_case(tournament.get("city") or ""),
        "country": tournament.get("country") or "",
        "drawSize": tournament.get("singlesDrawSize") or 0,
        "prize": tournament.get("prizeMoney") or 0,
        "currency": tournament.get("prizeMoneyCurrency") or "USD",
        "status": tournament.get("status") or "",
        "champion": (
            {
                "id": singles.get("id"),
                "name": singles.get("fullName"),
                "country": singles.get("countryCode") or "",
            }
            if singles
            else None
        ),
        "doublesChampions": [
            {"id": p.get("id"), "name": p.get("fullName"), "country": p.get("countryCode") or ""}
            for p in ((d or {}).get("player") for d in doubles)
            if p
        ],
    }


def main() -> int:
    log("tournaments", f"Fetching WTA Tour calendar for {PREV}–{YEAR}…")

    all_events: list[dict] = []
    for year in (PREV, YEAR):
        rows = collect(year)
        log("tournaments", f"{year}: {len(rows)} events")
        all_events.extend(rows)

    # De-duplicate on (group id, year, start date): multi-week events legitimately
    # appear once per week of play in the feed.
    seen: dict[str, dict] = {}
    for t in all_events:
        key = f"{(t.get('tournamentGroup') or {}).get('id')}|{t.get('year')}|{str(t.get('startDate') or '')[:10]}"
        seen.setdefault(key, t)

    events = [normalise(t) for t in seen.values()]
    events = [e for e in events if e["id"] and e["start"]]
    # Ascending then reversed: reverse=True would also reverse same-day events,
    # changing their order relative to the snapshot it replaces.
    events.sort(key=lambda e: e["start"])
    events.reverse()

    write_json(
        "data/tournaments.json",
        {
            "source": "WTA Official Tournaments API",
            "sourceUrl": "https://www.wtatennis.com/tournaments",
            "generatedAt": datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z"),
            "levels": LEVEL_ORDER,
            "seasons": [PREV, YEAR],
            "events": events,
        },
    )

    by_level: dict[str, int] = {}
    for event in events:
        by_level[event["level"]] = by_level.get(event["level"], 0) + 1
    log("tournaments", f"Done — {len(events)} events {by_level}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/fetch-tournaments.mjs
/**
 * Fetch the official WTA Tour calendar for the active seasons.
 *
 * The tournament feed carries the full draw metadata plus the champion of each
 * completed event, which is what the calendar view renders.
 */
import { fetchTournaments, fetchTournament, writeJson, log } from './lib.mjs';

const YEAR = Number(process.env.WTA_SEASON || new Date().getUTCFullYear());
const PREV = YEAR - 1;

log('tournaments', `Fetching WTA Tour calendar for ${PREV}–${YEAR}…`);

function range(year) {
  return { from: `${year}-01-01`, to: `${year}-12-31` };
}

/** Main-tour levels only: ITF World Tennis Tour events are out of scope. */
const TOUR_LEVELS = new Set([
  'Grand Slam',
  'WTA Finals',
  'WTA 1000',
  'WTA 500',
  'WTA 250',
  'WTA 125',
  'United Cup',
  'Billie Jean King Cup',
]);

async function collect(year) {
  const out = [];
  for (let page = 0; page < 12; page += 1) {
    const res = await fetchTournaments({
      page,
      pageSize: 100,
      ...range(year),
    });
    const content = res?.content;
    if (!Array.isArray(content) || content.length === 0) break;
    out.push(...content.filter((t) => TOUR_LEVELS.has(t?.level || t?.tournamentGroup?.level)));
    if (content.length < 100) break;
  }
  return out;
}

const LEVEL_ORDER = [
  'Grand Slam',
  'WTA Finals',
  'WTA 1000',
  'WTA 500',
  'WTA 250',
  'WTA 125',
  'ITF',
  'United Cup',
  'Billie Jean King Cup',
];

function normalise(t) {
  const group = t.tournamentGroup || {};
  const winner = t.winners?.[0];
  const singles = winner?.singles?.player;
  const doubles = Array.isArray(winner?.doubles) ? winner.doubles : [];
  return {
    id: group.id ?? null,
    name: titleCase(group.name || t.title || ''),
    title: t.title || '',
    year: t.year,
    level: t.level || group.level || '',
    start: (t.startDate || '').slice(0, 10),
    end: (t.endDate || '').slice(0, 10),
    surface: t.surface || '',
    indoor: t.inOutdoor === 'I',
    city: titleCase(t.city || ''),
    country: t.country || '',
    drawSize: t.singlesDrawSize || 0,
    prize: t.prizeMoney || 0,
    currency: t.prizeMoneyCurrency || 'USD',
    status: t.status || '',
    champion: singles
      ? { id: singles.id, name: singles.fullName, country: singles.countryCode || '' }
      : null,
    doublesChampions: doubles
      .map((d) => d?.player)
      .filter(Boolean)
      .map((p) => ({ id: p.id, name: p.fullName, country: p.countryCode || '' })),
  };
}

function titleCase(s) {
  return String(s)
    .toLowerCase()
    .split(/\s+/)
    .map((w) => (w ? w[0].toUpperCase() + w.slice(1) : w))
    .join(' ');
}

const all = [];
for (const year of [PREV, YEAR]) {
  const rows = await collect(year);
  log('tournaments', `${year}: ${rows.length} events`);
  all.push(...rows);
}

// De-duplicate on (group id, year, start date): multi-week events legitimately
// appear once per week of play in the feed.
const seen = new Map();
for (const t of all) {
  // `all` holds raw feed records here, so read the nested group id.
  const key = `${t?.tournamentGroup?.id}|${t?.year}|${(t?.startDate || '').slice(0, 10)}`;
  if (!seen.has(key)) seen.set(key, t);
}

const events = [...seen.values()]
  .map(normalise)
  .filter((t) => t.id && t.start) // keep only real, dated events
  .sort((a, b) => (a.start < b.start ? 1 : -1));

const payload = {
  source: 'WTA Official Tournaments API',
  sourceUrl: 'https://www.wtatennis.com/tournaments',
  generatedAt: new Date().toISOString(),
  levels: LEVEL_ORDER,
  seasons: [PREV, YEAR],
  events,
};

await writeJson('data/tournaments.json', payload);

const byLevel = {};
for (const e of events) byLevel[e.level] = (byLevel[e.level] || 0) + 1;
log('tournaments', `Done — ${events.length} events ${JSON.stringify(byLevel)}`);
